File: Iteration 3/Course.py
class Course:
    def __init__(self, courseId, courseName, credit, prerequisite, prerequisiteLessonId, courseSection,
                 weeklyCourseCount, year, instructor, enrollmentCapacity, currentCapacity, type, waitList=None):
        self.__courseId = courseId
        self.__courseName = courseName
        self.__credit = credit
        self.__prerequisite = prerequisite
        self.__prerequisiteLessonId = prerequisiteLessonId
        self.__courseSection = courseSection
        self.__weeklyCourseCount = weeklyCourseCount
        self.__year = year
        self.__instructor = instructor
        self.__enrollmentCapacity = enrollmentCapacity
        self.__currentCapacity = currentCapacity
        self.__type = type
        self.__waitList = waitList if waitList else []

    # Getter methods
    def get_course_id(self):
        return self.__courseId

    def get_course_name(self):
        return self.__courseName

    def get_credit(self):
        return self.__credit

    def get_prerequisite(self):
        return self.__prerequisite

    def get_prerequisite_lesson_id(self):
        return self.__prerequisiteLessonId

    def get_course_section(self):
        return self.__courseSection

    def get_weekly_course_count(self):
        return self.__weeklyCourseCount

    def get_year(self):
        return self.__year

    def get_instructor(self):
        return self.__instructor

    def get_enrollment_capacity(self):
        return self.__enrollmentCapacity

    def get_current_capacity(self):
        return self.__currentCapacity

    def get_type(self):
        return self.__type

    def get_wait_list(self):
        return self.__waitList

    # Setter methods
    def set_course_name(self, courseName):
        self.__courseName = courseName

    def set_credit(self, credit):
        self.__credit = credit

    def set_prerequisite(self, prerequisite):
        self.__prerequisite = prerequisite

    def set_prerequisite_lesson_id(self, prerequisiteLessonId):
        self.__prerequisiteLessonId = prerequisiteLessonId

    def set_course_section(self, courseSection):
        self.__courseSection = courseSection

    def set_weekly_course_count(self, weeklyCourseCount):
        self.__weeklyCourseCount = weeklyCourseCount

    def set_year(self, year):
        self.__year = year

    def set_instructor(self, instructor):
        self.__instructor = instructor

    def set_enrollment_capacity(self, enrollmentCapacity):
        self.__enrollmentCapacity = enrollmentCapacity

    def set_current_capacity(self, currentCapacity):
        self.__currentCapacity = currentCapacity

    def set_type(self, type):
        self.__type = type

    def set_wait_list(self, waitList):
        self.__waitList = waitList

    # Convert object to dictionary
    def to_dict(self):
        return {
            "courseId": self.__courseId,
            "courseName": self.__courseName,
            "credit": self.__credit,
            "prerequisite": self.__prerequisite,
            "prerequisiteLessonId": self.__prerequisiteLessonId,
            "courseSection": self.__courseSection,
            "weeklyCourseCount": self.__weeklyCourseCount,
            "year": self.__year,
            "instructor": self.__instructor,
            "enrollmentCapacity": self.__enrollmentCapacity,
            "currentCapacity": self.__currentCapacity,
            "type": self.__type,
            "waitList": self.__waitList
        }

File: Iteration2/CSE3063F24P1_GRP14/src/Course.java
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.ArrayList;
import java.util.List;

class Course {

    @JsonProperty("courseId")
    private String courseId;

    @JsonProperty("courseName")
    private String courseName;

    @JsonProperty("credit")
    private int credit;

    @JsonProperty("prerequisite")
    private boolean prerequisite;

    @JsonProperty("prerequisiteLessonId")
    private String prerequisiteLessonId;

    @JsonProperty("courseSection")
    private List<CourseSection> courseSection;

    @JsonProperty("weeklyCourseCount")
    private int weeklyCourseCount;

    @JsonProperty("year")
    private int year;

    @JsonProperty("instructor")
    private String instructor;

    @JsonProperty("enrollmentCapacity")
    private int enrollmentCapacity;

    @JsonProperty("currentCapacity")
    private int currentCapacity;

    @JsonProperty("status")
    private String status;

    @JsonProperty("waitList")
    private List<String> waitList;

    public Course() {}
    // Constructor
    public Course(String courseId, String courseName, int credit, boolean prerequisite, String prerequisiteLessonId,
                  List<CourseSection> courseSection, int weeklyCourseCount, int year, String instructor, int enrollmentCapacity, int currentCapacity, String status,
                  List<String> waitList) {
        this.courseId = courseId;
        this.courseName = courseName;
        this.credit = credit;
        this.prerequisite = prerequisite;
        this.prerequisiteLessonId = prerequisiteLessonId;
        this.courseSection = courseSection != null ? courseSection : new ArrayList<>();
        this.weeklyCourseCount = weeklyCourseCount;
        this.year = year;
        this.instructor = instructor;
        this.enrollmentCapacity = enrollmentCapacity;
        this.currentCapacity = currentCapacity;
        this.status = status;
        this.waitList = waitList != null ? waitList : new ArrayList<>();
    }

    public Course(String courseId, String courseName, int credit, boolean hasPrerequisite, String prerequisiteLessonId, List<CourseSection> updatedSections, int weeklyCourseCount, int year, String instructor, int enrollmentCapacity, int currentCapacity, String status) {
    }

//    public <E> Course(String c101, String testCourse, int i, boolean b, Object o, ArrayList<E> es, int i1, int i2, String s, int i3, int i4, String active) {
//    }

    // Getters
    public String getCourseId() {
        return courseId;
    }

    public String getCourseName() {
        return courseName;
    }

    public int getCredit() {
        return credit;
    }

    public boolean hasPrerequisite() {
        return prerequisite;
    }

    public String getPrerequisiteLessonId() {
        return prerequisiteLessonId;
    }

    public List<CourseSection> getCourseSection() {
        return courseSection;
    }

    public int getWeeklyCourseCount() {
        return weeklyCourseCount;
    }

    public int getYear() {
        return year;
    }

    public String getInstructor() {
        return instructor;
    }

    public int getEnrollmentCapacity() {
        return enrollmentCapacity;
    }

    public int getCurrentCapacity() {return currentCapacity;}

    public String getStatus() {
        return status;
    }

    public List<String> getWaitList(){return waitList;}

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Course course = (Course) obj;
        return courseId.equals(course.courseId); // Kursu `courseId` ile karşılaştırıyoruz
    }

    @Override
    public int hashCode() {
        return courseId.hashCode(); // `courseId`'yi hash değeri olarak kullanıyoruz
    }


    // toString method for printing
    @Override
    public String toString() {
        return "Course{" +
                "courseId='" + courseId + '\'' +
                ", courseName='" + courseName + '\'' +
                ", credit=" + credit +
                ", prerequisite=" + prerequisite +
                ", prerequisiteLessonId='" + prerequisiteLessonId + '\'' +
                ", courseSection=" + courseSection +
                ", weeklyCourseCount=" + weeklyCourseCount +
                ", year=" + year +
                ", instructor='" + instructor + '\'' +
                ", enrollmentCapacity=" + enrollmentCapacity +
                ", status='" + status + '\'' +
                ", waitList=" + waitList +
                '}';
    }


}